### ISSUE TITLE
Problem with cluster Configuration and error message MOVED 7250 <ipaddress>

### ISSUE THREAD

**Issue by jkljajic** [StackExchange/StackExchange.Redis#206]
I have big issue when i want to switch use of redis configuration for Cluster and almost every 5-6 response i get error MOVED .....

Looking in code for StackExchange.Redis i have found where is big problem and part 

in ResultProcessor.cs line 157

```C#
if (bridge.Multiplexer.TryResend(hashSlot, message, endpoint, isMoved))
{
    connection.Multiplexer.Trace(message.Command + " re-issued to " + endpoint,
        isMoved ? "MOVED" : "ASK");
    return false;
}
```

replace with simple 
(Problem is not here directly but this is solve my problem.)

return bridge.Multiplexer.TryResend(hashSlot, message, endpoint, isMoved); 
and all my problem gone.

i hope this can help some one or to you guys to solve issue with cluster configuration MOVED

**Comment by debbiswal** [comment]
Hi @jkljajic ,
Finally I got my program working. Thanks to @hrishi18pathak .
I have downloaded the code from https://github.com/hrishi18pathak/StackExchange.Redis/tree/SEClusteringBugFix branch with changes from 17-July-2015. The main changes are  in ConnectionMultiplexer and ConfigurartionOptions class.

********************************************************************************************
ConnectionMultiplexer.cs (modified UpdateClusterRange() method,added IsNodeCompatibleWithConfiguration() method) :
---------------------------------------------------------------------------------
 internal void UpdateClusterRange(ClusterConfiguration configuration)
        {
            if (configuration == null) return;
            foreach (var node in configuration.Nodes)
            {
                // do not update cluster range if the node configuration is incompatible with the multiplexer 
                // endpoints configuration. If a user has specified endpoints in the multiplexer configuration using
                // Dns endpoints, Stackexchange.redis returns the ConnectionMultiplexer object successfully 
                // after connecting to all the nodes in the connection string (using their host names). 
                // Also, before returning the multiplexer object it updates the mapping of nodes to their clusterslots 
                // and this mapping is of the format “DNS hostname:port” : “clusterSlotNumber”.
                // However at the same time, the client also issues “CLUSTER NODES” command to each of the nodes in the above list 
                // (to determine master and slave configuration) and re-updates the cluster mapping with the output of the clusternodes command 
                // which now is going to contain IP addresses (redis-server itself does not understand host names). 
                // So the cluster mapping is now updated to the format: “IP Address” : “clusterSlotNumber”
                // If the StackExchange.Redis has not been able to connect to all the nodes using their IP addresses by the time the first command (GET,SET) 
                // is issued to the cluster, it results in failure to connect to that particular node resulting in the “MOVED ERROR” 
                // (since it tries to hit a random node in the list subsequently)
                if (node.IsSlave || node.Slots.Count == 0 || !IsNodeCompatibleWithConfiguration(node)) continue;
                foreach (var slot in node.Slots)
                {
                    var server = GetServerEndPoint(node.EndPoint);
                    if (server != null) serverSelectionStrategy.UpdateClusterRange(slot.From, slot.To, server);
                }
            }
        }

 /// <summary>
        /// Checks if the specified node has the same format as that of
        /// the endpoints specified in the multiplexer configuration i.e.
        /// if all nodes are dns endpoints then the specified node has to be a dns endpoint
        /// to be compatible.
        /// </summary>
        /// <param name="node"></param>
        /// <returns>True if node is compatible with multiplexer configuration</returns>
        private bool IsNodeCompatibleWithConfiguration(ClusterNode node)
        {
            return (this.configuration.HasAllDnsEndPoints() && (node.EndPoint is DnsEndPoint)) ||
                (!this.configuration.HasDnsEndPoints() && !(node.EndPoint is DnsEndPoint));
        }
******************************************************************************************
ConfigurationOptions.cs(added method HasAllDnsEndPoints())
-----------------------------------------------------------------------------------
 internal bool HasAllDnsEndPoints()
        {
            return !endpoints.Any(ep => !(ep is DnsEndPoint));
        }

Still I need to verify for any implications with this fix.

**Comment by NickCraver** [comment]
This should be resolved in 1.1.553-alpha (latest), after #312 was merged in.

**Comment by NickCraver** [comment]
Merging into #248 for tracking of the MOVED exceptions.
